src/ir/auth_logic/ast.h

```diff
@@ -76,34 +78,35 @@ class Attribute {
 class CanActAs {
  public:
   explicit CanActAs(Principal left_principal, Principal right_principal)
-    : left_principal_(std::move(left_principal)),
-      right_principal_(std::move(right_principal)) {}
+      : left_principal_(std::move(left_principal)),
+        right_principal_(std::move(right_principal)) {}
   const Principal& left_principal() const { return left_principal_; }
   const Principal& right_principal() const { return right_principal_; }
 
  private:
-   Principal left_principal_;
-   Principal right_principal_;
+  Principal left_principal_;
+  Principal right_principal_;
 };
 
-// BaseFact corresponds to a base-case fact that cannot recursively include other 
-// facts through "canSay". "Fact" includes the recursive case. It is necessary 
-// that BaseFact is a separate class from Fact because the RHS of conditional 
-// assertions can only contain BaseFacts.
+// BaseFact corresponds to a base-case fact that cannot recursively include
+// other facts through "canSay". "Fact" includes the recursive case. It is
+// necessary that BaseFact is a separate class from Fact because the RHS of
+// conditional assertions can only contain BaseFacts.
 // Base facts have three forms:
 //  - A predicate
 //  - An attribute
 //  - A canActAs expression
-// The particular choice of form is represented with a variant which is 
-// encapsulated.
 class BaseFact {
  public:
-  explicit BaseFact(Predicate predicate) : value_(std::move(predicate)) {};
-  explicit BaseFact(Attribute attribute) : value_(std::move(attribute)) {};
-  explicit BaseFact(CanActAs canActAs) : value_(std::move(canActAs)) {};
+  // BaseFactVariantType gives the different forms of BaseFacts. Client code
+  // should use this type to traverse these forms. This type may be changed in
+  // the future.
+  using BaseFactVariantType = std::variant<Predicate, Attribute, CanActAs>;
```

Comment:
> How much do you rely on this in the public interface though?

The `BaseFactVariantType` would be used in all the passes that traverse the AST (right now this is `lowering_ast_datalog`).

This comment about how the public type alias should be used exists because the style guide says there should be such a comment: https://google.github.io/styleguide/cppguide.html#Aliases

> In any case, OK for now. However, it is better to get this resolved sooner rather than later before create too much dependency on this type.

I think making the variant accessible through a public accessor is OK (and it is not poor encapsulation). I think if the variant is publicly accessible it means the different variations of `BaseFactVariantType` are exposed in the API. This means the structure of the AST is visible in the API, but I think there is no way to hide the structure of the AST from the API. Even with a visitor, I think some of the passes that work with the visitor would need to change if the structure of the AST changes.

Using a visitor is definitely the more traditional way to implement this, but I think using `variant` and `std::visit` makes good use of the modern C++ features, and is in some ways improved from a conventional visitor (for example the visitors can have return types more easily).

Reply:
OK here is an issue [Visitor For Authorization Logic AST #345](https://github.com/google-research/raksha/issues/345)